sequencefile/reader.go

```diff
@@ -0,0 +1,177 @@
+// memory, not the value. Subsequent calls to Value will return nil.
+func (r *Reader) ScanKey() bool {
+	return r.scan(false)
+}
+
+// Err returns the first non-EOF error reached while scanning.
+func (r *Reader) Err() error {
+	return r.err
+}
+
+// Key returns the key for the current record. The byte slice will be reused
+// after the next call to Scan.
+func (r *Reader) Key() []byte {
+	return r.key
+}
+
+// Value returns the value for the current record. The byte slice will be
+// reused after the next call to Scan.
+func (r *Reader) Value() []byte {
+	return r.value
+}
+
+func (r *Reader) scan(readValues bool) bool {
+	if r.closed {
+		return false
+	}
+
+	var b []byte
+
+	b, err := r.consume(4)
+	if err == io.EOF {
+		return false
+	} else if err != nil {
+		r.close(err)
+		return false
+	}
+
+	// Length -1 means a sync marker (obnoxiously encoded as a cast uint32).
+	totalLength := int(int32(binary.BigEndian.Uint32(b)))
+	if totalLength == -1 {
+		return r.checkSyncAndScan(readValues)
+	} else if totalLength < 8 {
+		r.close(fmt.Errorf("Invalid record length: %d", totalLength))
+		return false
+	}
+
+	b, err = r.consume(4)
+	if err != nil {
+		r.close(err)
+		return false
+	}
+
+	totalKeyLength := int(int32(binary.BigEndian.Uint32(b)))
+	if totalKeyLength < 4 {
+		r.close(fmt.Errorf("Invalid key length: %d", totalKeyLength))
+	}
+
+	// Both the key and value have an extra 4 bytes of junk at the beginning for
+	// BytesWritable length.
+	keyStart := 4
+	keyEnd := keyStart + (totalKeyLength - 4)
+	valueStart := keyEnd + 4
+	valueEnd := valueStart + (totalLength - totalKeyLength - 4)
+
+	if readValues {
+		b, err = r.consume(totalLength)
+		if err != nil {
+			r.close(err)
+			return false
+		}
+
+		r.key = b[keyStart:keyEnd]
+		r.value = b[valueStart:valueEnd]
+	} else {
+		b, err = r.consume(totalKeyLength)
+		if err != nil {
+			r.close(err)
+			return false
+		}
+
+		// To discard the value, first we try seeking. If this isn't a file, then we
+		// just copy to /dev/null.
+		valueLength := int64(totalLength - totalKeyLength)
+		if seeker, ok := r.reader.(io.Seeker); ok {
+			_, err = seeker.Seek(valueLength, 1)
+		} else {
+			_, err = io.CopyN(ioutil.Discard, r.reader, valueLength)
+		}
+
+		if err != nil {
+			r.close(err)
+			return false
+		}
+
+		r.key = b[keyStart:keyEnd]
+		r.value = nil
+	}
+
+	return true
+}
+
+func (r *Reader) checkSyncAndScan(readValues bool) bool {
+	b, err := r.consume(SyncSize)
+	if err != nil {
+		r.close(err)
+		return false
+	}
+
+	// If we never read the Header, infer the sync marker from the first time we
+	// see it.
+	if r.syncMarkerBytes == nil {
+		r.syncMarkerBytes = b
+	} else if !bytes.Equal(b, r.syncMarkerBytes) {
+		r.close(fmt.Errorf("Invalid sync marker: %x", b))
+		return false
+	}
+
+	return r.scan(readValues)
+}
+
+func (r *Reader) consume(n int) ([]byte, error) {
+	r.buf.Reset()
+	r.buf.Grow(n)
+	b := r.buf.Bytes()[:n]
+
+	_, err := io.ReadFull(r.reader, b)
```

Comment:
As a naïve reader this looks to me like you're reusing the same buffer all the time. Wouldn't subsequent reads overwrite what you've previously read?